pkg/helm/renderer.go

```diff
@@ -0,0 +1,237 @@
+		Params: ast.Identifiers{"repository", "chart", "version", "params", "componentName"},
+		Func:   fn,
+	}
+
+	return nf
+}
+
+// Render renders a Helm chart.
+func (r *Renderer) Render(repoName, chartName, chartVersion, componentName string, values map[string]interface{}) ([]interface{}, error) {
+	logrus.WithFields(logrus.Fields{
+		"repoName":     repoName,
+		"chartName":    chartName,
+		"chartVersion": chartVersion,
+		"values":       values,
+	}).Debug("rendering helm chart")
+
+	if r.app == nil {
+		return nil, errors.New("app object is nil")
+	}
+
+	if chartVersion == "" {
+		var err error
+		chartVersion, err = LatestChartVersion(r.app, repoName, chartName)
+		if err != nil {
+			return nil, errors.Wrapf(err, "finding latest chart release for %s/%s", repoName, chartName)
+		}
+
+		logrus.WithFields(logrus.Fields{
+			"version":   chartVersion,
+			"chartName": chartName,
+			"repoName":  repoName}).Debug("using latest helm chart version because none was supplied")
+	}
+
+	chartPath := filepath.Join(r.app.Root(), "vendor", repoName, chartName, "helm", chartVersion, chartName)
+
+	b, err := goyaml.Marshal(values)
+	if err != nil {
+		return nil, err
+	}
+
+	rendered, err := r.renderWithHelm(componentName, string(b), chartPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "rendering Helm chart")
+	}
+
+	var out []interface{}
+	for name, s := range rendered {
+		if filepath.Ext(name) != ".yaml" {
+			continue
+		}
+		var m map[string]interface{}
+		if err := goyaml.Unmarshal([]byte(s), &m); err != nil {
+			return nil, errors.Wrapf(err, "unmarshalling %s", name)
+		}
+
+		out = append(out, m)
+	}
+
+	return out, nil
+}
+
+func (r *Renderer) renderWithHelm(componentName, raw, chartPath string) (map[string]string, error) {
+	config := &chart.Config{Raw: raw, Values: map[string]*chart.Value{}}
+
+	c, err := chartutil.LoadDir(chartPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "loading Helm chart")
+	}
+
+	if req, err := chartutil.LoadRequirements(c); err == nil {
+		if err := checkDependencies(c, req); err != nil {
+			return nil, err
+		}
+	} else if err != chartutil.ErrRequirementsNotFound {
+		return nil, fmt.Errorf("cannot load requirements: %v", err)
+	}
+
+	err = chartutil.ProcessRequirementsEnabled(c, config)
+	if err != nil {
+		return nil, err
+	}
+
+	err = chartutil.ProcessRequirementsImportValues(c)
+	if err != nil {
+		return nil, err
+	}
+
+	options := chartutil.ReleaseOptions{
+		Name: componentName,
+	}
+	vals, err := chartutil.ToRenderValuesCaps(c, config, options, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	renderer := engine.New()
+	rendered, err := renderer.Render(c, vals)
+	if err != nil {
+		return nil, errors.Wrap(err, "rendering Helm chart")
+	}
+
+	return rendered, nil
+}
+
+func checkDependencies(ch *chart.Chart, reqs *chartutil.Requirements) error {
+	missing := []string{}
+
+	deps := ch.GetDependencies()
+	for _, r := range reqs.Dependencies {
+		found := false
+		for _, d := range deps {
```

Comment:
What are the numbers we're dealing with here? Can we reimplement to be less O(n*m) as a followup issue?

Reply:
If it ever approached five, I'd be surprised.